File: chains/cpp/dexvm/src/manifest.cpp
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/dexvm/manifest.hpp"

#include "lux/dexvm/json.hpp"

#include <algorithm>
#include <array>
#include <cstdio>
#include <memory>
#include <set>

namespace lux::dexvm {
namespace {

// reject_unknown_fields is Go's DisallowUnknownFields, applied at every object a
// struct is decoded from. A typo'd field must be visible: a manifest that quietly
// dropped "asssets" would admit nothing and claim success.
Result<void> reject_unknown_fields(const json::Object& o, std::string_view what,
                                   std::span<const std::string_view> known) {
    for (const auto& [k, _] : o) {
        if (std::find(known.begin(), known.end(), k) == known.end())
            return fail("json: unknown field \"" + k + "\" in " + std::string(what));
    }
    return {};
}

const json::Value* member(const json::Object& o, std::string_view k) {
    auto it = o.find(std::string(k));
    if (it == o.end() || it->second.is_null()) return nullptr;
    return &it->second;
}

Result<std::uint64_t> u64_member(const json::Object& o, std::string_view k, std::uint64_t max) {
    const json::Value* v = member(o, k);
    if (!v) return std::uint64_t(0);
    auto n = v->as_u64(max);
    if (!n) return std::unexpected(n.error().wrap("field \"" + std::string(k) + "\""));
    return *n;
}

Result<std::string> string_member(const json::Object& o, std::string_view k) {
    const json::Value* v = member(o, k);
    if (!v) return std::string();
    auto s = v->as_string();
    if (!s) return std::unexpected(s.error().wrap("field \"" + std::string(k) + "\""));
    return *s;
}

Result<bool> bool_member(const json::Object& o, std::string_view k) {
    const json::Value* v = member(o, k);
    if (!v) return false;
    auto b = v->as_bool();
    if (!b) return std::unexpected(b.error().wrap("field \"" + std::string(k) + "\""));
    return *b;
}

Result<Id> id_member(const json::Object& o, std::string_view k) {
    const json::Value* v = member(o, k);
    if (!v) return kEmptyId;
    auto s = v->as_string();
    if (!s) return std::unexpected(s.error().wrap("field \"" + std::string(k) + "\""));
    auto id = id_from_string(*s);
    if (!id) return std::unexpected(id.error().wrap("field \"" + std::string(k) + "\""));
    return *id;
}

Result<Bytes> hex_member(const json::Object& o, std::string_view k) {
    const json::Value* v = member(o, k);
    if (!v) return Bytes{};
    auto s = v->as_string();
    if (!s) return std::unexpected(s.error().wrap("field \"" + std::string(k) + "\""));
    auto b = from_hex(*s);
    if (!b) return std::unexpected(b.error().wrap("field \"" + std::string(k) + "\""));
    return *b;
}

constexpr std::array<std::string_view, 9> kAssetFields{
    "networkID", "chainID", "assetKind", "canonicalRef", "decimals",
    "symbol",    "name",    "enabled",   "riskTier"};
constexpr std::array<std::string_view, 5> kMarketFields{"networkID", "baseAssetID", "quoteAssetID",
                                                        "venueConfig", "enabled"};
constexpr std::array<std::string_view, 7> kManifestFields{
    "network", "networkID", "evmChainID", "cChainID", "chainLabels", "assets", "markets"};

Result<Asset> decode_asset(const json::Value& v) {
    auto obj = v.as_object();
    if (!obj) return std::unexpected(obj.error());
    const json::Object& o = **obj;
    if (auto r = reject_unknown_fields(o, "registry.Asset", kAssetFields); !r)
        return std::unexpected(r.error());

    Asset a;
    auto net = u64_member(o, "networkID", 0xFFFFFFFFull);
    if (!net) return std::unexpected(net.error());
    a.network_id = std::uint32_t(*net);

    auto chain = id_member(o, "chainID");
    if (!chain) return std::unexpected(chain.error());
    a.chain_id = *chain;

    // assetKind decodes through the same closed-token parser the wire uses, so
    // an ASCII ticker in the kind slot fails here rather than downstream.
    if (const json::Value* kv = member(o, "assetKind")) {
        auto s = kv->as_string();
        if (!s) return std::unexpected(s.error().wrap("field \"assetKind\""));
        auto k = parse_kind(*s);
        if (!k) return std::unexpected(k.error());
        a.kind = *k;
    }

    auto ref = hex_member(o, "canonicalRef");
    if (!ref) return std::unexpected(ref.error());
    a.canonical_ref = *ref;

    auto dec = u64_member(o, "decimals", 0xFFull);
    if (!dec) return std::unexpected(dec.error());
    a.decimals = std::uint8_t(*dec);

    auto sym = string_member(o, "symbol");
    if (!sym) return std::unexpected(sym.error());
    a.symbol = *sym;

    auto nm = string_member(o, "name");
    if (!nm) return std::unexpected(nm.error());
    a.name = *nm;

    auto en = bool_member(o, "enabled");
    if (!en) return std::unexpected(en.error());
    a.enabled = *en;

    auto tier = u64_member(o, "riskTier", 0xFFull);
    if (!tier) return std::unexpected(tier.error());
    a.risk_tier = RiskTier(std::uint8_t(*tier));

    return a;
}

Result<Market> decode_market(const json::Value& v) {
    auto obj = v.as_object();
    if (!obj) return std::unexpected(obj.error());
    const json::Object& o = **obj;
    if (auto r = reject_unknown_fields(o, "registry.Market", kMarketFields); !r)
        return std::unexpected(r.error());

    Market m;
    auto net = u64_member(o, "networkID", 0xFFFFFFFFull);
    if (!net) return std::unexpected(net.error());
    m.network_id = std::uint32_t(*net);

    auto base = id_member(o, "baseAssetID");
    if (!base) return std::unexpected(base.error());
    m.base_asset_id = *base;

    auto quote = id_member(o, "quoteAssetID");
    if (!quote) return std::unexpected(quote.error());
    m.quote_asset_id = *quote;

    auto venue = hex_member(o, "venueConfig");
    if (!venue) return std::unexpected(venue.error());
    m.venue_config = *venue;

    auto en = bool_member(o, "enabled");
    if (!en) return std::unexpected(en.error());
    m.enabled = *en;

    return m;
}

void encode_asset(json::Writer& w, const Asset& a) {
    w.begin_object();
    w.key("networkID");
    w.number(a.network_id);
    w.key("chainID");
    w.string(cb58(a.chain_id));
    w.key("assetKind");
    w.string(to_string(a.kind));
    w.key("canonicalRef");
    w.string(hex0x(view(a.canonical_ref)));
    w.key("decimals");
    w.number(a.decimals);
    w.key("symbol");
    w.string(a.symbol);
    w.key("name");
    w.string(a.name);
    w.key("enabled");
    w.boolean(a.enabled);
    w.key("riskTier");
    w.number(std::uint8_t(a.risk_tier));
    w.end_object();
}

void encode_market(json::Writer& w, const Market& m) {
    w.begin_object();
    w.key("networkID");
    w.number(m.network_id);
    w.key("baseAssetID");
    w.string(cb58(m.base_asset_id));
    w.key("quoteAssetID");
    w.string(cb58(m.quote_asset_id));
    w.key("venueConfig");
    w.string(hex0x(view(m.venue_config)));
    w.key("enabled");
    w.boolean(m.enabled);
    w.end_object();
}

Result<Bytes> read_file(const std::string& path) {
    std::unique_ptr<std::FILE, int (*)(std::FILE*)> f(std::fopen(path.c_str(), "rb"), std::fclose);
    if (!f) return fail("manifest: read " + path + ": no such file or directory");
    Bytes out;
    std::array<std::uint8_t, 4096> buf{};
    while (true) {
        const std::size_t n = std::fread(buf.data(), 1, buf.size(), f.get());
        out.insert(out.end(), buf.begin(), buf.begin() + std::ptrdiff_t(n));
        if (n < buf.size()) break;
    }
    if (std::ferror(f.get())) return fail("manifest: read " + path + ": I/O error");
    return out;
}

}  // namespace

Result<void> Manifest::validate_shape() const {
    if (network.empty()) return fail("manifest: empty network name");
    if (network_id == 0) return fail("manifest: networkID must be non-zero");
    if (evm_chain_id == 0)
        return fail("manifest: evmChainID must be non-zero (the RPC-checkable C-Chain identity)");
    if (c_chain_id == kEmptyId)
        return fail("manifest: cChainID must be set (the C-Chain consensus id)");

    for (std::size_t i = 0; i < assets.size(); ++i) {
        const Asset& a = assets[i];
        const std::string at = "manifest: asset[" + std::to_string(i) + "]";
        if (a.network_id != network_id)
            return fail(at + " networkID " + std::to_string(a.network_id) +
                        " != manifest networkID " + std::to_string(network_id));
        switch (a.kind) {
            case AssetKind::EVMNative:
            case AssetKind::ERC20:
                if (a.chain_id != c_chain_id)
                    return fail(at + " (" + std::string(to_string(a.kind)) +
                                ") chainID must be the C-Chain " + cb58(c_chain_id) + ", got " +
                                cb58(a.chain_id));
                break;
            case AssetKind::UTXO:
                // Rooted at a UTXO source chain, not the C-Chain.
                break;
            default:
                return fail(at + " invalid kind");
        }
        if (auto r = a.validate_shape(); !r) return std::unexpected(r.error().wrap(at));
    }
    for (std::size_t i = 0; i < markets.size(); ++i) {
        if (markets[i].network_id != network_id)
            return fail("manifest: market[" + std::to_string(i) + "] networkID " +
                        std::to_string(markets[i].network_id) + " != manifest networkID " +
                        std::to_string(network_id));
    }
    return {};
}

std::function<std::string(const Id&)> Manifest::chain_label_for() const {
    // The labels are captured by value: the lookup outlives no manifest it did
    // not copy, so a gate run cannot read a freed map.
    return [labels = chain_labels](const Id& id) -> std::string {
        auto it = labels.find(hex(id));
        return it == labels.end() ? std::string() : it->second;
    };
}

Result<void> Manifest::admit_into(Registry& reg, ChainVerifier& v) const {
    if (auto r = validate_shape(); !r) return r;
    if (v.confirms_c_chain()) {
        if (auto r = v.confirm_c_chain(network_id, evm_chain_id, c_chain_id); !r)
            return std::unexpected(
                r.error().wrap("manifest " + network + ": C-Chain identity confirm"));
    }
    for (std::size_t i = 0; i < assets.size(); ++i) {
        if (auto r = reg.register_asset(assets[i], v); !r)
            return std::unexpected(r.error().wrap("manifest " + network + ": asset[" +
                                                  std::to_string(i) + "] (" +
                                                  std::string(to_string(assets[i].kind)) + ")"));
    }
    for (std::size_t i = 0; i < markets.size(); ++i) {
        if (auto r = reg.create_market(markets[i]); !r)
            return std::unexpected(
                r.error().wrap("manifest " + network + ": market[" + std::to_string(i) + "]"));
    }
    return {};
}

Result<void> Manifest::apply_to(Registry& reg, ChainVerifier& v,
                                const DexAssetPolicy& policy) const {
    if (auto r = admit_into(reg, v); !r) return r;
    const NetworkClass cls = network_class_for(network_id);
    if (auto r = refuse_under_synthetic_config(cls, policy, reg, chain_label_for()); !r)
        return std::unexpected(r.error().wrap("manifest " + network + ": startup gate"));
    return {};
}

Result<std::unique_ptr<Registry>> Manifest::validate(ChainVerifier& v) const {
    auto reg = std::make_unique<Registry>(
        std::vector<AssetKind>{AssetKind::EVMNative, AssetKind::ERC20, AssetKind::UTXO});
    if (auto r = apply_to(*reg, v, default_dex_asset_policy()); !r)
        return std::unexpected(r.error());
    return reg;
}

std::string Manifest::encode() const {
    json::Writer w;
    w.begin_object();
    w.key("network");
    w.string(network);
    w.key("networkID");
    w.number(network_id);
    w.key("evmChainID");
    w.number(evm_chain_id);
    w.key("cChainID");
    w.string(cb58(c_chain_id));
    // chainLabels carries omitempty, so an absent or empty map is not written at
    // all. Its keys are written sorted, which is what Go does for every map.
    if (!chain_labels.empty()) {
        w.key("chainLabels");
        w.begin_object();
        for (const auto& [k, v] : chain_labels) {
            w.key(k);
            w.string(v);
        }
        w.end_object();
    }
    w.key("assets");
    if (assets_is_null) {
        w.null();
    } else {
        w.begin_array();
        for (const Asset& a : assets) encode_asset(w, a);
        w.end_array();
    }
    w.key("markets");
    if (markets_is_null) {
        w.null();
    } else {
        w.begin_array();
        for (const Market& m : markets) encode_market(w, m);
        w.end_array();
    }
    w.end_object();
    return w.str();
}

Result<DecodedManifest> decode_manifest_bytes(ByteView raw, std::string_view label) {
    DecodedManifest out;
    out.sha256_hex = hex(view(sha256(raw)));

    auto doc = json::parse(std::string_view(reinterpret_cast<const char*>(raw.data()), raw.size()));
    if (!doc)
        return std::unexpected(doc.error().wrap("manifest: decode " + std::string(label)));
    auto obj = doc->as_object();
    if (!obj)
        return std::unexpected(obj.error().wrap("manifest: decode " + std::string(label)));
    const json::Object& o = **obj;
    if (auto r = reject_unknown_fields(o, "registry.Manifest", kManifestFields); !r)
        return std::unexpected(r.error().wrap("manifest: decode " + std::string(label)));

    Manifest& m = out.manifest;
    auto ctx = [&](Error e) { return e.wrap("manifest: decode " + std::string(label)); };

    auto network = string_member(o, "network");
    if (!network) return std::unexpected(ctx(network.error()));
    m.network = *network;

    auto net = u64_member(o, "networkID", 0xFFFFFFFFull);
    if (!net) return std::unexpected(ctx(net.error()));
    m.network_id = std::uint32_t(*net);

    auto evm = u64_member(o, "evmChainID", 0xFFFFFFFFFFFFFFFFull);
    if (!evm) return std::unexpected(ctx(evm.error()));
    m.evm_chain_id = *evm;

    auto cchain = id_member(o, "cChainID");
    if (!cchain) return std::unexpected(ctx(cchain.error()));
    m.c_chain_id = *cchain;

    if (const json::Value* lv = member(o, "chainLabels")) {
        auto lo = lv->as_object();
        if (!lo) return std::unexpected(ctx(lo.error().wrap("field \"chainLabels\"")));
        for (const auto& [k, v] : **lo) {
            auto s = v.as_string();
            if (!s) return std::unexpected(ctx(s.error().wrap("chainLabels[\"" + k + "\"]")));
            m.chain_labels[k] = *s;
        }
    }

    if (const json::Value* av = member(o, "assets")) {
        auto arr = av->as_array();
        if (!arr) return std::unexpected(ctx(arr.error().wrap("field \"assets\"")));
        m.assets_is_null = false;
        for (std::size_t i = 0; i < (*arr)->size(); ++i) {
            auto a = decode_asset((**arr)[i]);
            if (!a)
                return std::unexpected(
                    ctx(a.error().wrap("assets[" + std::to_string(i) + "]")));
            m.assets.push_back(std::move(*a));
        }
    }

    if (const json::Value* mv = member(o, "markets")) {
        auto arr = mv->as_array();
        if (!arr) return std::unexpected(ctx(arr.error().wrap("field \"markets\"")));
        m.markets_is_null = false;
        for (std::size_t i = 0; i < (*arr)->size(); ++i) {
            auto mk = decode_market((**arr)[i]);
            if (!mk)
                return std::unexpected(
                    ctx(mk.error().wrap("markets[" + std::to_string(i) + "]")));
            m.markets.push_back(std::move(*mk));
        }
    }

    if (auto r = m.validate_shape(); !r)
        return std::unexpected(r.error().wrap("manifest " + std::string(label)));
    return out;
}

Result<Manifest> load_manifest(const std::string& path) {
    auto raw = read_file(path);
    if (!raw) return std::unexpected(raw.error());
    auto d = decode_manifest_bytes(view(*raw), path);
    if (!d) return std::unexpected(d.error());
    return std::move(d->manifest);
}

Result<std::string> normalize_sha256(std::string_view h) {
    std::string s;
    for (char c : h) {
        if (c == ' ' || c == '\t' || c == '\n' || c == '\r') continue;
        s.push_back(char(std::tolower(static_cast<unsigned char>(c))));
    }
    if (s.starts_with("sha256:")) s.erase(0, 7);
    if (s.starts_with("0x")) s.erase(0, 2);
    if (s.size() != 64)
        return fail("registry: pinned manifest hash must be a 32-byte SHA-256 (64 hex chars), got " +
                    std::to_string(s.size()) + " chars");
    for (char c : s) {
        if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')))
            return fail("registry: pinned manifest hash is not valid hex");
    }
    return s;
}

Result<Manifest> load_manifest_pinned(const std::string& path, std::string_view expected_sha256) {
    if (expected_sha256.empty()) return load_manifest(path);
    auto want = normalize_sha256(expected_sha256);
    if (!want) return std::unexpected(want.error());
    auto raw = read_file(path);
    if (!raw) return std::unexpected(raw.error());
    auto d = decode_manifest_bytes(view(*raw), path);
    if (!d) return std::unexpected(d.error());
    if (d->sha256_hex != *want)
        return fail_note(Err::ManifestHashMismatch,
                         "path=" + path + " want=" + *want + " got=" + d->sha256_hex);
    return std::move(d->manifest);
}

}  // namespace lux::dexvm

File: chains/rust/dexvm/src/manifest.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! The per-network declaration of what trades.
//!
//! There is exactly one manifest per network. Every entry must be real on the
//! named network; CI proves this against the network's RPC before any deploy,
//! and the node re-proves the identity binding at startup.
//!
//! The manifest does NOT carry derived AssetIDs or MarketIDs — those are
//! computed from the canonical fields, so the file cannot disagree with the
//! identity.
//!
//! A manifest's bytes are its own pin. The content SHA-256 is taken over the
//! EXACT bytes read, so a pinned-hash check binds the loaded manifest to the
//! CI-approved artifact byte-for-byte. That is what stops a locally edited file
//! — a fabricated token address — from being loaded by a node that holds no EVM
//! state of its own and so cannot look the token up itself.

use crate::asset::AssetKind;
use crate::error::{Code, Error, Result};
use crate::gate::{network_class_for, refuse_under_synthetic_config, DexAssetPolicy};
use crate::ids::{self, Id};
use crate::market::Market;
use crate::registry::{Asset, ChainVerifier, Registry};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// The on-disk declaration of the REAL assets and markets the DEX admits on one
/// network.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, default)]
pub struct Manifest {
    /// The canonical network name. It must match the deploy target; a mismatch
    /// is a hard error — you cannot ship the testnet manifest to mainnet.
    pub network: String,
    /// The Lux networkID every asset and market in this manifest must declare.
    #[serde(rename = "networkID")]
    pub network_id: u32,
    /// The C-Chain's EVM chainID (`eth_chainId`). It is the AUTHORITATIVE,
    /// RPC-checkable identity of the C-Chain: the CI validator confirms the
    /// target RPC reports this before admitting any ERC-20 or native entry, so
    /// a manifest can never be validated against the wrong chain.
    #[serde(rename = "evmChainID")]
    pub evm_chain_id: u64,
    /// The canonical C-Chain CONSENSUS id, used in the AssetID preimage so a
    /// derived AssetID lives in the same identity space as the on-chain atomic
    /// objects. EVM_NATIVE and ERC20 entries are rooted here.
    #[serde(rename = "cChainID", with = "crate::ids::serde_cb58")]
    pub c_chain_id: Id,
    /// Source chain id (as HEX, which is how Go keys it) to human label,
    /// consumed by the forbidden-reference deny-scan. Optional; an unlabeled
    /// chain simply has no white-label name to match.
    #[serde(
        rename = "chainLabels",
        skip_serializing_if = "Option::is_none",
        default
    )]
    pub chain_labels: Option<BTreeMap<String, String>>,
    /// The declared real entries.
    pub assets: Vec<Asset>,
    pub markets: Vec<Market>,
}

impl Default for Manifest {
    fn default() -> Self {
        Manifest {
            network: String::new(),
            network_id: 0,
            evm_chain_id: 0,
            c_chain_id: ids::EMPTY,
            chain_labels: None,
            assets: Vec::new(),
            markets: Vec::new(),
        }
    }
}

impl Manifest {
    /// Whether the manifest is internally consistent, before any chain I/O:
    /// network name present, every asset and market on the manifest's network,
    /// every C-Chain asset rooted at the manifest's C-Chain, and every asset
    /// structurally valid.
    pub(crate) fn validate_shape(&self) -> Result<()> {
        if self.network.is_empty() {
            return Err(Error::other("manifest: empty network name"));
        }
        if self.network_id == 0 {
            return Err(Error::other("manifest: networkID must be non-zero"));
        }
        if self.evm_chain_id == 0 {
            return Err(Error::other(
                "manifest: evmChainID must be non-zero (the RPC-checkable C-Chain identity)",
            ));
        }
        if ids::is_empty(&self.c_chain_id) {
            return Err(Error::other(
                "manifest: cChainID must be set (the C-Chain consensus id)",
            ));
        }
        for (i, a) in self.assets.iter().enumerate() {
            if a.network_id != self.network_id {
                return Err(Error::other(format!(
                    "manifest: asset[{i}] networkID {} != manifest networkID {}",
                    a.network_id, self.network_id
                )));
            }
            match a.kind {
                AssetKind::EvmNative | AssetKind::Erc20 => {
                    if a.chain_id != self.c_chain_id {
                        return Err(Error::other(format!(
                            "manifest: asset[{i}] ({}) chainID must be the C-Chain {}, got {}",
                            a.kind,
                            ids::cb58(&self.c_chain_id),
                            ids::cb58(&a.chain_id)
                        )));
                    }
                }
                // A UTXO asset is rooted at a UTXO source chain, not the C-Chain.
                AssetKind::Utxo => {}
                AssetKind::Invalid => {
                    return Err(Error::other(format!("manifest: asset[{i}] invalid kind")))
                }
            }
            a.validate_shape()
                .map_err(|e| e.wrap(format!("manifest: asset[{i}]")))?;
        }
        for (i, m) in self.markets.iter().enumerate() {
            if m.network_id != self.network_id {
                return Err(Error::other(format!(
                    "manifest: market[{i}] networkID {} != manifest networkID {}",
                    m.network_id, self.network_id
                )));
            }
        }
        Ok(())
    }

    /// The deny-scan label lookup for this manifest's declared chain labels.
    /// The map is keyed by an id's HEX, which is what Go writes there — reading
    /// it as cb58 would find nothing and silently pass every scan.
    pub fn chain_label_for(&self) -> impl Fn(Id) -> String + '_ {
        move |id: Id| match &self.chain_labels {
            None => String::new(),
            Some(m) => m.get(&ids::hex(&id)).cloned().unwrap_or_default(),
        }
    }

    /// Register every manifest asset (each proven real) and create every
    /// manifest market (each pinned to two registered assets), WITHOUT running
    /// the startup gate.
    ///
    /// It is the admission half of [`Manifest::apply_to`], separated so a caller
    /// that owns the gate — the node, which runs it once with its own network
    /// class and policy — does not run it twice.
    pub fn admit_into(&self, reg: &Registry, v: &dyn ChainVerifier) -> Result<()> {
        self.validate_shape()?;
        v.confirm_c_chain(self.network_id, self.evm_chain_id, self.c_chain_id)
            .map_err(|e| {
                e.wrap(format!(
                    "manifest {}: C-Chain identity confirm",
                    self.network
                ))
            })?;
        for (i, a) in self.assets.iter().enumerate() {
            reg.register(a, v).map_err(|e| {
                e.wrap(format!(
                    "manifest {}: asset[{i}] ({})",
                    self.network, a.kind
                ))
            })?;
        }
        for (i, m) in self.markets.iter().enumerate() {
            reg.create_market(m)
                .map_err(|e| e.wrap(format!("manifest {}: market[{i}]", self.network)))?;
        }
        Ok(())
    }

    /// Admit everything, then run the fail-closed startup gate for this
    /// manifest's network class and the given policy.
    ///
    /// This is the ONE routine that turns a manifest into a live, gated registry
    /// — used identically by CI with an RPC verifier and by node startup with
    /// the local-chain verifier. Any failure aborts; `reg` is left partially
    /// populated only on error, and callers discard it.
    pub fn apply_to(
        &self,
        reg: &Registry,
        v: &dyn ChainVerifier,
        policy: &DexAssetPolicy,
    ) -> Result<()> {
        self.admit_into(reg, v)?;
        let class = network_class_for(self.network_id);
        let label = self.chain_label_for();
        refuse_under_synthetic_config(class, policy, reg, Some(&label))
            .map_err(|e| e.wrap(format!("manifest {}: startup gate", self.network)))
    }

    /// The full CI check for one manifest against one verifier: prove every
    /// entry real and gate-clean in a fresh registry under the canonical
    /// locked-down policy. The populated registry is returned so a caller can
    /// report what was admitted.
    pub fn validate(&self, v: &dyn ChainVerifier) -> Result<Registry> {
        let reg = Registry::new(&[AssetKind::EvmNative, AssetKind::Erc20, AssetKind::Utxo]);
        self.apply_to(&reg, v, &DexAssetPolicy::default_policy())?;
        Ok(reg)
    }
}

/// Content-hash and decode raw manifest bytes — from a file or from the
/// binary's own embedded copy — returning the manifest and the bytes' SHA-256 in
/// lowercase hex.
///
/// This is the SINGLE manifest decoder, so shape validation, unknown-field
/// rejection and content-hash discipline are identical for both. `label` is
/// error context only: a path, or an embed name.
pub fn decode_manifest_bytes(raw: &[u8], label: &str) -> (String, Result<Manifest>) {
    let hex_sum = ids::hex(&ids::sha256(raw));
    // A typo'd field (`asssets`) fails closed — `deny_unknown_fields` on every
    // struct in the tree is what Go's DisallowUnknownFields does for the whole
    // decode.
    let decoded: std::result::Result<Manifest, _> = serde_json::from_slice(raw);
    let m = match decoded {
        Ok(m) => m,
        Err(e) => {
            return (
                hex_sum,
                Err(Error::other(format!("manifest: decode {label}: {e}"))),
            )
        }
    };
    if let Err(e) = m.validate_shape() {
        return (hex_sum, Err(e.wrap(format!("manifest {label}"))));
    }
    (hex_sum, Ok(m))
}

/// Read a manifest file and its content hash.
fn load_manifest_bytes(path: &std::path::Path) -> (String, Result<Manifest>) {
    let raw = match std::fs::read(path) {
        Ok(r) => r,
        Err(e) => {
            return (
                String::new(),
                Err(Error::other(format!(
                    "manifest: read {}: {e}",
                    path.display()
                ))),
            )
        }
    };
    decode_manifest_bytes(&raw, &path.display().to_string())
}

/// Read and decode a manifest file. It does NOT verify against chain state —
/// that is [`Manifest::apply_to`], which needs a verifier — only parse and
/// structurally validate. A malformed kind, ref or tier fails here.
pub fn load_manifest(path: &std::path::Path) -> Result<Manifest> {
    load_manifest_bytes(path).1
}

/// Read a manifest and REFUSE it unless its content SHA-256 equals
/// `expected_sha256` (lowercase hex, with or without a `0x` or `sha256:`
/// prefix).
///
/// An empty expectation means "no pin configured" and falls back to
/// [`load_manifest`] — pinning is opt-in per deployment, but once a hash is set
/// the file must match it exactly. A malformed expectation is itself an error:
/// you cannot pin against garbage.
pub fn load_manifest_pinned(path: &std::path::Path, expected_sha256: &str) -> Result<Manifest> {
    if expected_sha256.is_empty() {
        return load_manifest(path);
    }
    let want = normalize_sha256(expected_sha256)?;
    let (got, m) = load_manifest_bytes(path);
    let m = m?;
    if got != want {
        return Err(Error::note(
            Code::ManifestHashMismatch,
            format!("path={} want={want} got={got}", path.display()),
        ));
    }
    Ok(m)
}

/// Lowercase, strip an optional `0x` or `sha256:` prefix, and check the value is
/// a 64-hex-character digest. A non-conforming value fails closed.
pub fn normalize_sha256(h: &str) -> Result<String> {
    let s = h.trim().to_lowercase();
    let s = s.strip_prefix("sha256:").unwrap_or(&s);
    let s = s.strip_prefix("0x").unwrap_or(s);
    if s.len() != 64 {
        return Err(Error::other(format!(
            "registry: pinned manifest hash must be a 32-byte SHA-256 (64 hex chars), got {} chars",
            s.len()
        )));
    }
    if !s.bytes().all(|c| c.is_ascii_hexdigit()) {
        return Err(Error::other(
            "registry: pinned manifest hash is not valid hex",
        ));
    }
    Ok(s.to_string())
}
